### PR TITLE
Use offset index in ParquetRecordBatchStream

### LINKED ISSUE #2430: Support Reading PageIndex with ParquetRecordBatchStream
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`ParquetRecordBatchStream` does not contain the necessary logic to read the offset index

**Describe the solution you'd like**

Add support for reading the column index to `ParquetRecordBatchStream`

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #2434: Error Reading Page Index When Not Available
**Describe the bug**

The doc comment for `read_pages_locations` reads

> If not the format not available return an empty vector.

I would take this to mean it would return an empty vector if index information is not available, instead it returns

```
General("underlying Thrift error: end of file")
```

**To Reproduce**

Enable `ReadOptions::enable_page_index` and try to read a parquet file that lacks a PageIndex

**Expected behavior**

Enabling the page index shouldn't require it to exist

**Additional context**

### PR DESCRIPTION
**Pull request by thinkharderdev**
# Which issue does this PR close?

Closes #2430 
Closes #2434 

# Rationale for this change
 

Leverage OffsetIndex (if available) to prune IO in ParquetRecordBatchStream

# What changes are included in this PR?

Allow user to specify read options in ParquetRecordBatchStreamBuilder which will fetch index metadata when building. 

Assorted bug fixes:
* Bug in `RowSelection::scan_ranges` when skipping past the final page boundary
* We were not fetching dictionary pages when creating `SerializedPageReader` from `InMemoryColumnChunk` if we had a page index. 

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/file/page_index/index_reader.rs [hunk 1]
```diff
@@ -34,8 +34,12 @@ pub fn read_columns_indexes<R: ChunkReader>(
     let (offset, lengths) = get_index_offset_and_lengths(chunks)?;
     let length = lengths.iter().sum::<usize>();
 
+    if length == 0 {
```
Comment (thinkharderdev) [inline parquet/src/file/page_index/index_reader.rs:37]: Not sure if this is right or we should return an empty vec?

File: parquet/src/arrow/async_reader.rs [hunk 2]
```diff
@@ -139,6 +143,80 @@ pub trait AsyncFileReader: Send {
     /// allowing fine-grained control over how metadata is sourced, in particular allowing
     /// for caching, pre-fetching, catalog metadata, etc...
     fn get_metadata(&mut self) -> BoxFuture<'_, Result<Arc<ParquetMetaData>>>;
+
+    /// Provides asynchronous access to the the page index for each column chunk in a
+    /// row group. Will panic if `row_group_idx` is greater than or equal to `num_row_groups`
+    fn get_column_indexes(
```
Comment (thinkharderdev) [inline parquet/src/arrow/async_reader.rs:149]: Not sure if we should make this a separate constructor or just have it always use the index. If the index doesn't exist, the cost of determining that is minimal.

File: object_store/src/local.rs [hunk 3]
```diff
@@ -1068,6 +1068,7 @@ mod tests {
         integration.head(&path).await.unwrap();
     }
 
+    #[ignore]
```
Comment (tustvold) [inline object_store/src/local.rs:1071]: :smile:

File: parquet/src/arrow/async_reader.rs [hunk 4]
```diff
@@ -218,6 +296,36 @@ impl<T: AsyncFileReader + Send + 'static> ArrowReaderBuilder<AsyncReader<T>> {
         Self::new_builder(AsyncReader(input), metadata, Default::default())
     }
 
+    pub async fn new_with_index(mut input: T) -> Result<Self> {
```
Comment (tustvold) [inline parquet/src/arrow/async_reader.rs:299]: I think it would be more consistent to have `new_with_options` accepting `ArrowReaderOptions`. This already has a field on it for the page index

File: parquet/src/arrow/async_reader.rs [hunk 5]
```diff
@@ -218,6 +296,36 @@ impl<T: AsyncFileReader + Send + 'static> ArrowReaderBuilder<AsyncReader<T>> {
...
+            columns_indexes.push(column_index);
+            if let Some(offset_index) =
+                input.get_page_locations(metadata.clone(), idx).await?
+            {
+                rg.set_page_offset(offset_index.clone());
+                offset_indexes.push(offset_index);
+            }
+        }
+
+        let metadata = Arc::new(ParquetMetaData::new_with_page_index(
```
Comment (tustvold) [inline parquet/src/arrow/async_reader.rs:319]: Not part of this PR, but I still feel something is off with the way the index information is located on ParquetMetadata...

File: parquet/src/file/page_index/index_reader.rs [hunk 6]
```diff
@@ -64,6 +68,10 @@ pub fn read_pages_locations<R: ChunkReader>(
 ) -> Result<Vec<Vec<PageLocation>>, ParquetError> {
     let (offset, total_length) = get_location_offset_and_total_length(chunks)?;
 
+    if total_length == 0 {
```
Comment (tustvold) [inline parquet/src/file/page_index/index_reader.rs:71]: I think this might fix https://github.com/apache/arrow-rs/issues/2434

File: parquet/src/arrow/arrow_reader/selection.rs [hunk 7]
```diff
@@ -162,7 +162,12 @@ impl RowSelection {
                     current_selector = selectors.next();
                 }
             } else {
-                break;
+                if !(selector.skip || current_page_included) {
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/selection.rs:165]: What are the implications of this change?

File: parquet/src/arrow/async_reader.rs [hunk 8]
```diff
@@ -139,6 +143,80 @@ pub trait AsyncFileReader: Send {
...
+            let (offset, lengths) = index_reader::get_index_offset_and_lengths(chunks)?;
+            let length = lengths.iter().sum::<usize>();
+
+            if length == 0 {
+                return Ok(vec![Index::NONE; chunks.len()]);
+            }
+
+            //read all need data into buffer
+            let data = self
+                .get_bytes(offset as usize..offset as usize + length)
```
Comment (tustvold) [inline parquet/src/arrow/async_reader.rs:166]: This will perform separate get_bytes requests to fetch the page index and column index information for each column chunk. This is likely not a good idea, especially since this will be performed serially. 

Ideally we would identify the ranges of all the index information, and then call get_byte_ranges, this will allow coalescing proximate requests to ObjectStore, paralell fetch, etc...

File: parquet/src/arrow/async_reader.rs [hunk 9]
```diff
@@ -218,6 +296,36 @@ impl<T: AsyncFileReader + Send + 'static> ArrowReaderBuilder<AsyncReader<T>> {
...
+    pub async fn new_with_index(mut input: T) -> Result<Self> {
+        let metadata = input.get_metadata().await?;
+
+        let mut row_groups = metadata.row_groups().to_vec();
+
+        let mut columns_indexes = vec![];
+        let mut offset_indexes = vec![];
+
+        for (idx, rg) in row_groups.iter_mut().enumerate() {
+            let column_index = input.get_column_indexes(metadata.clone(), idx).await?;
```
Comment (tustvold) [inline parquet/src/arrow/async_reader.rs:308]: We should check if the column index has already been fetched to the metadata, and not fetch it again if it is already present

File: parquet/src/arrow/async_reader.rs (hunk 2 above)
Comment (tustvold) [reply parquet/src/arrow/async_reader.rs:149]: I'm not sure of the value of exposing these on `AsyncFileReader`, and not just handling the logic internally. Ultimately if the implementer wants to override the way the index is fetched, they can just return ParquetMetadata from `get_metadata` with the index information already loaded.

**Review by tustvold — COMMENTED** [review]
I think I would opt to keep more of this logic hidden, i.e. not exposed on `AsyncFileReader`, and make use `get_byte_ranges` to avoid making lots of separate small fetch requests

File: parquet/src/arrow/async_reader.rs (hunk 5 above)
Comment (thinkharderdev) [reply parquet/src/arrow/async_reader.rs:319]: Yeah, I agree.

File: parquet/src/arrow/arrow_reader/selection.rs (hunk 7 above)
Comment (thinkharderdev) [reply parquet/src/arrow/arrow_reader/selection.rs:165]: Before we would break if we were on the last page. So if you skipped from inside the second to last page into the last page, then this would short circuit and the last page range wouldn't be selected.

File: parquet/src/arrow/async_reader.rs (hunk 8 above)
Comment (thinkharderdev) [reply parquet/src/arrow/async_reader.rs:166]: Yeah, that makes sense.

File: parquet/src/arrow/async_reader.rs [hunk 10]
```diff
@@ -493,13 +601,26 @@ impl<'a> InMemoryRowGroup<'a> {
...
                 .into_iter()
-                .filter_map(|(idx, chunk)| {
+                .filter_map(|(idx, (chunk, chunk_meta))| {
                     (chunk.is_none() && projection.leaf_included(idx)).then(|| {
-                        let ranges = selection.scan_ranges(&page_locations[idx]);
+                        // If the first page does not start at the beginning of the column,
+                        // then we need to also fetch a dictionary page.
+                        let mut ranges = vec![];
+                        let (start, _len) = chunk_meta.byte_range();
+                        match page_locations[idx].first() {
```
Comment (thinkharderdev) [inline parquet/src/arrow/async_reader.rs:613]: Discovered this bug as well. We weren't fetching a dictionary page if it existed

File: parquet/src/column/reader.rs [hunk 11]
```diff
@@ -312,7 +312,7 @@ where
 
                 // If page has less rows than the remaining records to
                 // be skipped, skip entire page
-                if metadata.num_rows < remaining {
+                if metadata.num_rows <= remaining {
```
Comment (thinkharderdev) [inline parquet/src/column/reader.rs:315]: Another bug I uncovered while testing. This causes the reader to try and fetch the pages unnecessarily which are not pre-fetched

File: parquet/src/arrow/async_reader.rs [hunk 12]
```diff
@@ -218,6 +221,96 @@ impl<T: AsyncFileReader + Send + 'static> ArrowReaderBuilder<AsyncReader<T>> {
...
+                let (idx_offset, idx_lengths) =
+                    index_reader::get_index_offset_and_lengths(rg.columns())?;
+                let idx_length = idx_lengths.iter().sum::<usize>();
+
+                // If index data is missing, return without any indexes
+                if loc_length == 0 || idx_length == 0 {
+                    return Self::new_builder(AsyncReader(input), metadata, options);
+                }
+
+                fetch_ranges.push(loc_offset as usize..loc_offset as usize + loc_length);
```
Comment (Ted-Jiang) [inline parquet/src/arrow/async_reader.rs:252]: I think this will read one col_index and page_location alternately. but they are written separately.
https://github.com/apache/parquet-format/blob/master/doc/images/PageIndexLayout.png

I think if we not cache all bytes in memory, we should read whole col_index then page_location.
why not we use 🤔
```rust
/// Read on row group's all columns indexes and change into  [`Index`]
/// If not the format not available return an empty vector.
pub fn read_columns_indexes<R: ChunkReader>(
    reader: &R,
    chunks: &[ColumnChunkMetaData],
) -> Result<Vec<Index>, ParquetError> {
```

Comment (tustvold) [reply parquet/src/arrow/async_reader.rs:252]: This can't use read_columns_indexes as this needs to asynchronously fetch the bytes

File: parquet/src/arrow/async_reader.rs [hunk 13]
```diff
@@ -218,6 +221,96 @@ impl<T: AsyncFileReader + Send + 'static> ArrowReaderBuilder<AsyncReader<T>> {
...
+        {
+            let mut fetch_ranges = vec![];
+            let mut index_lengths: Vec<Vec<usize>> = vec![];
+
+            for rg in metadata.row_groups() {
+                let (loc_offset, loc_length) =
+                    index_reader::get_location_offset_and_total_length(rg.columns())?;
+
+                let (idx_offset, idx_lengths) =
+                    index_reader::get_index_offset_and_lengths(rg.columns())?;
```
Comment (tustvold) [inline parquet/src/arrow/async_reader.rs:244]: It occurs to me that this method is making a pretty strong assumption that the column index data is contiguous, I'm not sure this is actually guaranteed...

File: parquet/src/arrow/async_reader.rs [hunk 14]
```diff
@@ -218,6 +221,96 @@ impl<T: AsyncFileReader + Send + 'static> ArrowReaderBuilder<AsyncReader<T>> {
...
+                let mut offset_index = vec![];
+
+                for _ in 0..columns.len() {
+                    let mut prot = TCompactInputProtocol::new(&mut cursor);
+                    let offset = OffsetIndex::read_from_in_protocol(&mut prot)?;
+                    offset_index.push(offset.page_locations);
+                }
+
+                rg.set_page_offset(offset_index.clone());
+                offset_indexes.push(offset_index);
```
Comment (tustvold) [inline parquet/src/arrow/async_reader.rs:279]: Again this interface seems really confused - something for #2530

File: parquet/src/arrow/async_reader.rs (hunk 12 above)
Comment (Ted-Jiang) [reply parquet/src/arrow/async_reader.rs:252]: Oh! Thanks! but we can still try combine all `col_index` or `page_location` together separately

**Review by tustvold — APPROVED** [review]
Looks good to me, thank you. There is definitely some cleanup to do with the page index plumbing, but that is beyond the scope of this PR.

File: parquet/src/arrow/async_reader.rs (hunk 12 above)
Comment (tustvold) [reply parquet/src/arrow/async_reader.rs:252]: The onus is on `AsyncFileReader::get_ranges` to handle this, e.g. `ObjectStore::get_ranges` does this already
